File: Services/PathfindingService/Models/Ray.cs
namespace VMAP
{
    /// <summary>
    /// Ray with origin and direction (direction is normalized).
    /// </summary>
    public struct Ray
    {
        private Vector3 origin;
        private Vector3 direction;

        public Ray(Vector3 origin, Vector3 direction, bool normalizeDir = true)
        {
            this.origin = origin;
            if (normalizeDir)
                this.direction = direction.Normalized();
            else
                this.direction = direction;
        }

        public Vector3 Origin => origin;
        public Vector3 Direction => direction;
        /// <summary>Component-wise reciprocal of the direction (with Inf for 0).</summary>
        public readonly Vector3 invDirection()
        {
            static float inv(float v) => MathF.Abs(v) < 1e-8f ? float.PositiveInfinity : 1f / v;
            return new Vector3(inv(Direction.x), inv(Direction.y), inv(Direction.z));
        }
        /// <summary>
        /// Computes the intersection distance of this ray with an axis-aligned bounding box.
        /// Returns positive infinity if no intersection within [0, +inf) range.
        /// </summary>
        public float IntersectionTime(AABox box)
        {
            // Use slabs method for Ray-AABB intersection
            float tMin = 0.0f;
            float tMax = float.PositiveInfinity;

            // For each axis, compute intersection t range
            // Direction components could be 0 (parallel to axis).
            if (MathF.Abs(direction.x) < 1e-8f)
            {
                if (origin.x < box.Min.x || origin.x > box.Max.x)
                    return float.PositiveInfinity; // no hit (ray is parallel and outside slab)
            }
            else
            {
                float invDir = 1.0f / direction.x;
                float t1 = (box.Min.x - origin.x) * invDir;
                float t2 = (box.Max.x - origin.x) * invDir;
                if (t1 > t2) (t1, t2) = (t2, t1); // swap
                tMin = MathF.Max(tMin, t1);
                tMax = MathF.Min(tMax, t2);
                if (tMin > tMax) return float.PositiveInfinity;
            }
            if (MathF.Abs(direction.y) < 1e-8f)
            {
                if (origin.y < box.Min.y || origin.y > box.Max.y)
                    return float.PositiveInfinity;
            }
            else
            {
                float invDir = 1.0f / direction.y;
                float t1 = (box.Min.y - origin.y) * invDir;
                float t2 = (box.Max.y - origin.y) * invDir;
                if (t1 > t2) (t1, t2) = (t2, t1);
                tMin = MathF.Max(tMin, t1);
                tMax = MathF.Min(tMax, t2);
                if (tMin > tMax) return float.PositiveInfinity;
            }
            if (MathF.Abs(direction.z) < 1e-8f)
            {
                if (origin.z < box.Min.z || origin.z > box.Max.z)
                    return float.PositiveInfinity;
            }
            else
            {
                float invDir = 1.0f / direction.z;
                float t1 = (box.Min.z - origin.z) * invDir;
                float t2 = (box.Max.z - origin.z) * invDir;
                if (t1 > t2) (t1, t2) = (t2, t1);
                tMin = MathF.Max(tMin, t1);
                tMax = MathF.Min(tMax, t2);
                if (tMin > tMax) return float.PositiveInfinity;
            }
            // If we reach here, the ray intersects the box between tMin and tMax.
            if (tMin < 0f)
            {
                // Ray origin is inside the box or just at boundary; treat as hit at t=0
                tMin = 0f;
            }
            return tMin;
        }
    }
}

File: Exports/Navigation/Ray.cpp
// Ray.cpp
#include "Ray.h"
#include <algorithm>
#include "VMapLog.h"

namespace G3D
{
    Ray::Ray()
    {
    }

    Ray::Ray(const Vector3& org, const Vector3& dir)
        : m_origin(org), m_direction(dir)
    {
        m_invDirection = Vector3(
            dir.x != 0 ? 1.0f / dir.x : inf(),
            dir.y != 0 ? 1.0f / dir.y : inf(),
            dir.z != 0 ? 1.0f / dir.z : inf()
        );
    }

    float Ray::intersectionTime(const AABox& box) const
    {
        const float EPSILON = 1e-6f;  // Add epsilon for stability

        // Log ray and box details
        LOG_DEBUG("=== Ray::intersectionTime ===");
        LOG_DEBUG("Ray origin: (" << m_origin.x << ", " << m_origin.y << ", " << m_origin.z << ")");
        LOG_DEBUG("Ray direction: (" << m_direction.x << ", " << m_direction.y << ", " << m_direction.z << ")");
        LOG_DEBUG("Box min: (" << box.low().x << ", " << box.low().y << ", " << box.low().z << ")");
        LOG_DEBUG("Box max: (" << box.high().x << ", " << box.high().y << ", " << box.high().z << ")");

        float tmin = -inf();
        float tmax = inf();

        for (int i = 0; i < 3; ++i)
        {
            const char* axisName = (i == 0) ? "X" : (i == 1) ? "Y" : "Z";

            // Check for parallel ray (near-zero direction component)
            if (std::abs(m_direction[i]) < EPSILON)
            {
                LOG_DEBUG("  " << axisName << "-axis: Ray parallel to slab");
                LOG_DEBUG("    Origin[" << i << "]=" << m_origin[i]
                    << " Box range=[" << box.low()[i] << ", " << box.high()[i] << "]");

                // Ray is parallel to slab, check if origin is within slab
                if (m_origin[i] < box.low()[i] || m_origin[i] > box.high()[i])
                {
                    LOG_DEBUG("    Ray origin outside slab bounds - NO INTERSECTION");
                    return inf();
                }
                LOG_DEBUG("    Ray origin within slab bounds - continuing");
            }
            else
            {
                float t1 = (box.low()[i] - m_origin[i]) * m_invDirection[i];
                float t2 = (box.high()[i] - m_origin[i]) * m_invDirection[i];

                LOG_DEBUG("  " << axisName << "-axis calculations:");
                LOG_DEBUG("    invDirection[" << i << "]=" << m_invDirection[i]);
                LOG_DEBUG("    t1 (to low)=" << t1 << " t2 (to high)=" << t2);

                // Handle negative direction
                if (m_invDirection[i] < 0.0f)
                {
                    std::swap(t1, t2);
                    LOG_DEBUG("    Swapped due to negative direction: t1=" << t1 << " t2=" << t2);
                }

                float old_tmin = tmin;
                float old_tmax = tmax;
                tmin = std::max(tmin, t1);
                tmax = std::min(tmax, t2);

                LOG_DEBUG("    Updated: tmin " << old_tmin << " -> " << tmin
                    << ", tmax " << old_tmax << " -> " << tmax);

                // Early exit if no intersection
                if (tmin > tmax)
                {
                    LOG_DEBUG("    tmin > tmax (" << tmin << " > " << tmax << ") - NO INTERSECTION");
                    return inf();
                }
            }
        }

        // Return the entry point (or 0 if ray starts inside)
        float result = tmin > 0 ? tmin : 0;
        LOG_DEBUG("Final result: tmin=" << tmin << " returning " << result
            << (result == 0 ? " (ray starts inside box)" : " (intersection distance)"));

        return result;
    }

    Ray Ray::fromOriginAndDirection(const Vector3& org, const Vector3& dir)
    {
        return Ray(org, dir);
    }
}